components/tidb_query/src/rpn_expr/impl_math.rs

```diff
@@ -183,42 +251,135 @@ impl Floor for FloorIntToInt {
     }
 }
 
-#[rpn_fn]
 #[inline]
-fn abs_int(arg: &Option<Int>) -> Result<Option<Int>> {
-    match arg {
-        None => Ok(None),
-        Some(arg) => match (*arg).checked_abs() {
-            None => Err(Error::overflow("BIGINT", &format!("abs({})", *arg)).into()),
-            Some(arg_abs) => Ok(Some(arg_abs)),
-        },
+#[rpn_fn(capture = [ctx])]
+pub fn round<R: Round>(ctx: &mut EvalContext, arg: &Option<R::Type>) -> Result<Option<R::Type>> {
+    if let Some(arg) = arg {
+        R::round(ctx, arg)
+    } else {
+        Ok(None)
     }
 }
 
-#[rpn_fn]
 #[inline]
-fn abs_uint(arg: &Option<Int>) -> Result<Option<Int>> {
-    Ok(*arg)
+#[rpn_fn(capture = [ctx])]
+pub fn round_with_frac<R: Round>(
+    ctx: &mut EvalContext,
+    arg0: &Option<R::Type>,
+    arg1: &Option<Int>,
+) -> Result<Option<R::Type>> {
+    match (arg0, arg1) {
+        (Some(number), Some(digits)) => R::round_with_frac(ctx, number, *digits),
+        _ => Ok(None),
+    }
 }
 
-#[rpn_fn]
-#[inline]
-fn abs_real(arg: &Option<Real>) -> Result<Option<Real>> {
-    match arg {
-        Some(arg) => Ok(Some(num_traits::Signed::abs(arg))),
-        None => Ok(None),
+pub trait Round {
+    type Type: Evaluable;
+
+    fn round(_ctx: &mut EvalContext, _arg: &Self::Type) -> Result<Option<Self::Type>> {
+        Ok(None)
+    }
+    fn round_with_frac(
+        _ctx: &mut EvalContext,
+        _number: &Self::Type,
+        _digits: Int,
+    ) -> Result<Option<Self::Type>> {
+        Ok(None)
     }
 }
 
-#[rpn_fn]
-#[inline]
-fn abs_decimal(arg: &Option<Decimal>) -> Result<Option<Decimal>> {
-    match arg {
-        Some(arg) => {
-            let res: codec::Result<Decimal> = arg.to_owned().abs().into();
-            Ok(Some(res?))
+pub struct RoundInt;
+
+impl Round for RoundInt {
+    type Type = Int;
+
+    #[inline]
+    fn round(_ctx: &mut EvalContext, arg: &Self::Type) -> Result<Option<Self::Type>> {
+        Ok(Some(*arg))
+    }
+}
+
+pub struct RoundReal;
+
+impl Round for RoundReal {
+    type Type = Real;
+
+    #[inline]
+    fn round(_ctx: &mut EvalContext, arg: &Self::Type) -> Result<Option<Self::Type>> {
+        Ok(Some(Real::from(arg.round())))
```

Comment:
Prefer to use `Real::new()` everywhere in non-test code since it won't panic for NaNs.

Reply:
Okay